### PR TITLE
Refactor e2e & Joules Plugin

### PR DESCRIPTION
**Pull request by nayiri-k**
This PR adds many more capabilities to the e2e tests, so that we can easily run flows for different designs/pdks/tools. The e2e/README summarizes how to run this flow. The PR also adds several features to the Cadence Joules power plugin, including writing database checkpoints between power steps, and supporting several reporting output formats (notably power profile instead of average power reporting).

**Related PRs / Issues**

**Type of change**:
- [ ] Bug fix
- [x] New feature
- [x] Other enhancement

**Impact**:
- [x] Change to core Hammer
- [x] Change to a Hammer plugin
- [ ] Other

**Contributor Checklist**:
- [x] Did you set `master` as the base branch?
- [x] Did you state the type-of-change/impact?
- [ ] Did you delete any extraneous prints/debugging code?
- [x] (If applicable) Did you add documentation for the feature?
- [ ] (If applicable) Did you update the `poetry.lock` file if you updated the requirements in `pyproject.toml`?
- [ ] (If applicable) Did you add a unit test demonstrating the PR?
- [ ] (If applicable) Did you run this through the e2e integration tests?
- [ ] (If applicable) Did you update the submodules in `e2e/` if this feature depends on updated plugins?

### CODE REVIEW

**Review by harrisonliew — CHANGES_REQUESTED** [review]
Looks great! Some global comments:
 

1. With these updates to the README, I think we can directly include it into the readthedocs (instead of reference). See this folder for an example for how to symlink the Markdown file: https://github.com/ucb-bar/hammer/tree/master/doc/Technology.
2. `test` seems like it should be called `pass`? Perhaps this directory should actually be using the `nop` tools and techs so that it can truly be used for testing w/o any plugin dependencies. Then the power ymls should apply to GCD instead.

File: hammer/power/joules/defaults.yml [hunk 1]
```diff
@@ -7,5 +7,5 @@ power.joules:
   joules_bin_meta: lazysubst # we want later overrides to be able to affect this
 
   # Joules version to use.
-  # Used to locate the binary - e.g. the '201' in ${cadence.cadence_home}/JLS/JLS201/bin/joules
-  version: "201"
+  # Used to locate the binary - e.g. the '211' in ${cadence.cadence_home}/JLS/JLS211/bin/joules
+  version: "211"
```
Comment (harrisonliew) [inline hammer/power/joules/defaults.yml:11]: Can you try using 22.1? It is also now merged under DDI (like #798)

File: e2e/configs-design/gcd/sky130.yml [hunk 2]
```diff
@@ -46,11 +51,3 @@ vlsi.inputs.pin.generate_mode: semi_auto
 vlsi.inputs.pin.assignments: [
   {pins: "*", layers: ["met4"], side: "bottom"}
 ]
-
```
Comment (dpgrubb13) [inline e2e/configs-design/gcd/sky130.yml:49]: Can you add timing constraints for the GCD module here?

File: e2e/configs-design/gcd/asap7.yml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    x: 0
+    y: 0
+    width: 100
+    height: 100
+    margins:
+      left: 0
+      right: 0
+      top: 0
+      bottom: 0
+
```
Comment (dpgrubb13) [inline e2e/configs-design/gcd/asap7.yml:23]: Also add timing constraints on the I/O here.

File: e2e/configs-design/gcd/sim-par.yml [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+sim.inputs:
+  top_module: gcd
+  tb_dut: gcd_dut
+  tb_name: "gcd_tb"
+  input_files: ['src/gcd_tb.v']
+  input_files_meta: append
+  level: "par"
+  options: ['-timescale=1ns/10ps']
+  options_meta: append
+  timing_annotated: false
```
[the remark this answers is not in the document]
Comment (nayiri-k) [reply e2e/configs-design/gcd/sim-par.yml]: This was copied from the previous `e2e` files. Just changed to `true` & tested.

**Review by harrisonliew — APPROVED** [review]

File: e2e/configs-design/gcd/common.yml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+# Generate Make include to aid in tlow
```
[the remark this answers is not in the document]
Comment (dpgrubb13) [reply e2e/configs-design/gcd/common.yml:1]: Follow up on the flow nit here.

File: e2e/configs-design/pass/joules.yml [hunk 6]
```diff
@@ -3,9 +3,4 @@
 vlsi.core.power_tool: "hammer.power.joules"
 
 power.inputs.input_files: ["/path/to/pass.v"]
-power.inputs.top_module: "pass"
-power.inputs.tb_name: "pass_tb"
-power.inputs.tb_dut: "pass_0"
 power.inputs.waveforms: ["path/to/waveforem"]
```
Comment (dpgrubb13) [inline e2e/configs-design/pass/joules.yml:6]: Nit: waveform

**Review by dpgrubb13 — APPROVED** [review]
Just change the few nits. Looks great!

**Pull request merged**